Content.Server/Fluids/EntitySystems/AbsorbentSystem.cs

```diff
@@ -185,13 +198,39 @@ private bool TryPuddleInteract(EntityUid user, EntityUid used, EntityUid target,
             return true;
         }
 
-        var transferMax = absorber.PickupAmount;
-        var transferAmount = available > transferMax ? transferMax : available;
+        // hack to compensate loss due to fractions that would be periodical
+        // and make sure (available / Ratio) * Ratio is not less than available
+        // otherwise the absorber would very annoyingly never fill
+        // it's not necessary if the ratio is not periodical in a decimal base
+        if (PuddleSystem.EvaporationReagentRatio % 2 != 0 &&
+                PuddleSystem.EvaporationReagentRatio % 5 != 0){
+            available += FixedPoint2.Epsilon;
+        }
+
+        var dirtVolumeAvailable = puddleSolution.Volume - puddleSolution.GetReagentQuantity(PuddleSystem.EvaporationReagent);
+
+        var transferAmount = FixedPoint2.Min(dirtVolumeAvailable,
+                                             available * PuddleSystem.EvaporationReagentRatio,
+                                             absorber.PickupAmount);
+
+        var prospectWaterTransfer = FixedPoint2.Min(transferAmount / PuddleSystem.EvaporationReagentRatio,
+                                                    absorberSoln.GetReagentQuantity(PuddleSystem.EvaporationReagent));
+
+        // This var exists for no other purpose than code readability
+        var prospectAvailableVolume = absorberSoln.AvailableVolume + prospectWaterTransfer;
```

Comment:
This code looks pretty hacky, the whole reason to use fixedpoint2 is to not get a loss of precision.

I think you should just adjust the transfer amount to the puddle by multiplying it by the evaporation amount as some of the other changes look like they cause issues with mopping.

Reply:
You can't use division in scalars in computers and not have loss of precision, that's just a limitation from the machine itself. The only way to have exact divisions is using a more symbolic division operation, such as by not representing the result as a scalar number, but as a rational number, and store a pair of integers as dividend and divisor. But then that only works for integer division. Fixed point is very useful but it doesn't make inherently inexact operations into exact.

If you want to ascertain this by yourself just take any calculator and divide a number by 9. Nine is cool cause it's not divisible by either 2 or 5, so it ends up causing recurring decimals when represented in a decimal base - or binary. The only cases of exact numeric division by 9 after being represented as a scalar (in a base which is not 3) is when the dividend is a multiple of 9. This stands for any digital computer. Feel free to ping me on discord about any of this.

Please name any changes you think cause issue with mopping so I can investigate. We have been running this code since I made it without any complaints for now.

Reply:
You only see a relevant loss of precision with ints or FixedPoint2 when you do a later multiplication. So the trick is to turn code 

```
a /= 10
a *= 20
```

Into the mathematically equivalent (but precision preserving)

```
a *= 20
a /= 10
```

i.e you do things which make your scalars large up front so they have precision going into the division.

Reply:
this is correct, but clearly not helpful at all since these operations happen at different interactions in this code. I can't simply make the mop clean first then be filled after cleaning.

Reply:
this comment here also points out: It has to remove water (division) prior to checking if there's space for puddle volume
`+        var water = refillableSolution.RemoveReagent(PuddleSystem.EvaporationReagent, transferAmount);
+
+        // Reads from refillableSolution.AvailableVolume have to
+        // be done after possibly removing some water
+        if (refillableSolution.AvailableVolume == FixedPoint2.Zero){
`